File: Data Structures & Algorithms/string-encode-and-decode/submission-1.go
type Solution struct{}

func (s *Solution) Encode(strs []string) string {
    var b strings.Builder

    for _, s := range strs {
        b.WriteString(strconv.Itoa(len(s)))
        b.WriteByte('#')
        b.WriteString(s)
    }
    return b.String()
}

func (s *Solution) Decode(encoded string) []string {
    res := []string{}

    i := 0

    for i < len(encoded) {
        j := i

        for encoded[j] != '#' {
            j++
        }

        length, _ := strconv.Atoi(encoded[i:j])

        start := j + 1
        end := start + length

        res = append(res, encoded[start:end])

        i = end
    }
    return res
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        l, r, res = [1] * n, [1] * n , [1] * n

        l[0] = 1
        r[n-1] = 1

        for i in range(1, n):
            l[i] = l[i-1] * nums[i-1]

        res[n-1] = l[n-1]  * r[n-1]

        for i in range(n-2, -1, -1):
            r[i] = r[i+1] * nums[i+1]
            res[i] = l[i] * r[i]
        return res